### ISSUE TITLE
"java.lang.SecurityException: Prohibited package name: java.io" when redefining classes from JDK

### ISSUE THREAD

**Issue by geld0r** [raphw/byte-buddy#133]
I'am trying to intercept & redefine classes from JDK (e.g. FileInputStream) which - if I understood correctly - should be possible with ByteBuddy. In the corresponding test however I receive the above error. My test is based on this projects tests and uses a ByteArrayClassLoader. I am not sure if this is causing the error of this also comes up in production yet.
A similar test for a custom class works without said exception.
If required I can provide the code that produced the error, but I guess this is easy to reproduce.

See detail description of issue: [https://stackoverflow.com/questions/37399087/bytebuddy-how-to-test-with-classes-from-java](StackOverflow)

**Comment by raphw** [comment]
The `java.*` namespace is treated specially by the VM. It is only possible to load such classes by the VM's bootstrap class loader. This means that you need to use a Java agent to redefine this class instead of loading the class by another class loader (what is forbidden).

This also means that you need to redefine the class what limits the scope of the possible transformations.

**Comment by geld0r** [comment]
Can you be a bit more specific please?
What exactly do I need to do differently in a test for my agent as opposed to the example tests?
Is there any way at all to make sure that the agent is attached before the class is loaded?
Currently I am attaching the agent in the tests @Before method.
In case (Unit-)Testing this is not possible at all, how do you suggest to approach developing an agent?

**Comment by raphw** [comment]
The Byte Buddy unit tests load a class in a dedicated, newly created class loader in order to make sure that the class in question is not yet loaded when the Java agent is attached. This way, the Java agent can assume that it is instrumenting the class in question before it was loaded for the first time. Otherwise, the unit tests would need to assume that no other test has loaded the class in question previously and introduce a test-order-dependency.

This approach is unfortunately not applicable for `java.*` classes. These classes **must** be loaded by the bootstrap class loader, this restriction is hard-coded into the `ClassLoader` implementation. It might just be that the `FileInputStream` class is not loaded by the bootstrap loader when you execute your test, but you cannot safely assume that.

Generally, I would be careful to instrument `java.*` classes with an `Instrumentation` as it is perfectly possible that for example `FileInputStream` is loaded before you have the chance to register an agent. The class is for example used by Byte Buddy itself at multiple points and you would actively need to avoid these code segments in order to make your instrumentation work. Alternatively, have a look at the `Advice` component which only adds code to the beginning and start of a method without changing the class file format. Such code changes are allowed even after a class was already loaded. This is also possible to unit test without assuming that the `FileInputStream` class was not yet loaded.

**Comment by geld0r** [comment]
Thank you for explanation.
What I want to achieve is "only" to be notified whenever a FileInputStream (and potentially similar streams/files/handles) are opened and closed so that I can keep track of this (leak detection)
Given your explanation, this probably can't be achieved reliably with my current approach.
In that case I'll try out `Advice`, although I am not sure if I understood how to use that yet.

**Comment by raphw** [comment]
With the `Advice` component, you can add code before or after a method without changing the layout of the class (adding or removing methods/fields). An advice for a file input stream could look like:

```java
class BeforeAdvice {
    @Advice.OnMethodEnter
    static void enter() {
      System.out.println("Will be printed before adviced method is called");
    }
}
```

Using `Advice.to(BeforeAdvice.class).on(matcher())` now allows you to add this code to the beginning of a method. Make sure to activate the `disableClassFormatChanges()` option on the `AgentBuilder`. You can subsequently activate the `with(RedefinitionStrategy.RETRANSFORM)` on the agent and redefine the `FileInputStream` class even after it was loaded.

**Comment by geld0r** [comment]
Thank you, this works fine and is exactly what I was looking for!

I've run into another error though that I don't understand:
I am getting a java.lang.NoClassDefFoundError when the Advice is executed and refers to any other class (e.g. logger or custom classes). When I comment this out, the error doesn't occur anymore.

This is the advice:
```
@Advice.OnMethodExit
public static void exit(@Advice.Argument(0) File file)
{
    System.out.println("Advice out: " +file);
    RecordFactory.getInstance().create(file);
}
```
I'm using the following AgentBuilder:
```
new AgentBuilder.Default()
				.disableClassFormatChanges()
				.with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
				.with(new LoggingListener())
				.type(ElementMatchers.isSubTypeOf(FileInputStream.class))
				.transform(((builder, typeDescription, classLoader) ->                                               builder.visit(Advice.to(BeforeAdvice.class).on(matcher))))
				.installOnByteBuddyAgent();
```

I've already tried out the various intialization strategies without success.

**Comment by raphw** [comment]
This is not a Byte Buddy specific problem. If you instrument a class, that class is bound by the visibility of its `ClassLoader`. The bootstrap class loader only knows classes that are defined by the JVM. You can however add jar files using the `Instrumentation` interface via `ByteBuddyAgent.getInstrumentation().appendToBootstrapSearch( ... )` or using injection via `ClassInjector.ForInstrumentation`.

**Comment by geld0r** [comment]
I've tried the `ByteBuddyAgent.getInstrumentation().appendToBootstrapSearch( ... )` approach. This really seems to rely on the code being present in the form of a jar file? At least when running from an IDE this is not the case. I've now tried to split the project into several modules so that I can load the relevant classes from a prebuilt jar.
Is there any easier way to achieve this?
I've not been able to get it running with the other approach.
